File: lstore/transaction_worker.py
import threading
import time
from lstore.table import Table, Record
from lstore.index import Index

class TransactionWorker:
    """
    # Creates a transaction worker object.
    """
    def __init__(self, transactions = None):
        self.stats = []
        self.transactions = transactions if transactions is not None else []
        self.result = 0
        self.thread = None  # Store the worker's thread
    
    """
    Append t to transactions
    """
    def add_transaction(self, t):
        self.transactions.append(t)
        
    """
    Run all transaction as a thread
    """
    def run(self):
        # Create a thread that executes __run
        self.thread = threading.Thread(target=self.__run)
        self.thread.start()
    
    """
    Wait for the worker to finish
    """
    def join(self):
        if self.thread:
            self.thread.join()

    def __run(self):
        MAX_RETRIES = 100
        
        # Process each transaction in the batch
        for transaction in self.transactions:
            retry_count = 0
            
            # Retry loop with exponential backoff
            while retry_count < MAX_RETRIES:
                result = transaction.run()
                
                # Transaction succeeded
                if result:
                    self.stats.append(True)
                    self.result += 1
                    break
                
                # Transaction aborted - backoff before retry
                retry_count += 1
                delay = 0.001 * min(retry_count, 10)  # 10ms delay
                jitter = delay * 0.5 * (hash(threading.current_thread()) % 100) / 100
                time.sleep(delay + jitter)
            
            # Max retries exceeded
            else:
                self.stats.append(False)

File: lstore/java-db/src/main/java/com/dylangutierrez/lstore/TransactionWorker.java
package com.dylangutierrez.lstore;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.LockSupport;

/**
 * Runs a batch of transactions on a dedicated thread with retries + backoff.
 * Contract: for each transaction, record exactly one boolean in stats:
 *   true  = succeeded within maxRetries
 *   false = failed (max retries exceeded or interrupted)
 */
public final class TransactionWorker {

    @FunctionalInterface
    public interface TransactionTask {
        boolean run();
    }

    @FunctionalInterface
    public interface Sleeper {
        void sleepNanos(long nanos) throws InterruptedException;
    }

    private static final Sleeper DEFAULT_SLEEPER = nanos -> LockSupport.parkNanos(nanos);

    private final List<TransactionTask> transactions;
    private final List<Boolean> stats; // one entry per transaction
    private final AtomicInteger result;

    private final int maxRetries;
    private final Sleeper sleeper;

    private Thread thread;

    public TransactionWorker() {
        this(null, 100, DEFAULT_SLEEPER);
    }

    public TransactionWorker(List<TransactionTask> transactions) {
        this(transactions, 100, DEFAULT_SLEEPER);
    }

    public TransactionWorker(List<TransactionTask> transactions, int maxRetries, Sleeper sleeper) {
        if (maxRetries <= 0) {
            throw new IllegalArgumentException("maxRetries must be > 0");
        }
        this.transactions = new ArrayList<>();
        if (transactions != null) {
            this.transactions.addAll(transactions);
        }
        this.maxRetries = maxRetries;
        this.sleeper = Objects.requireNonNull(sleeper, "sleeper");
        this.stats = Collections.synchronizedList(new ArrayList<>());
        this.result = new AtomicInteger(0);
        this.thread = null;
    }

    public void addTransaction(TransactionTask t) {
        Objects.requireNonNull(t, "transaction");
        this.transactions.add(t);
    }

    public void run() {
        if (this.thread != null) {
            throw new IllegalStateException("TransactionWorker already started");
        }
        this.thread = new Thread(this::internalRun, "transaction-worker");
        this.thread.start();
    }

    public void join() throws InterruptedException {
        if (this.thread != null) {
            this.thread.join();
        }
    }

    public int getResult() {
        return result.get();
    }

    public List<Boolean> getStatsSnapshot() {
        synchronized (stats) {
            return List.copyOf(stats);
        }
    }

    private void internalRun() {
        for (TransactionTask transaction : transactions) {
            boolean success = false;

            for (int attempt = 0; attempt < maxRetries; attempt++) {
                try {
                    success = transaction.run();
                } catch (RuntimeException ex) {
                    // Treat unexpected runtime errors as a failed attempt (retryable)
                    success = false;
                }

                if (success) {
                    break;
                }

                // If this was the last attempt, don't bother sleeping/backing off
                if (attempt == maxRetries - 1) {
                    break;
                }

                // Exponential-ish backoff like Python version:
                // baseDelay = 1ms * min(attempt+1, 10), plus jitter up to 50%
                int retryCount = attempt + 1;
                long baseDelayMillis = Math.min(retryCount, 10);
                long baseDelayNanos = TimeUnit.MILLISECONDS.toNanos(baseDelayMillis);

                double jitterFrac = ThreadLocalRandom.current().nextDouble(0.0, 0.5);
                long jitterNanos = (long) (baseDelayNanos * jitterFrac);

                try {
                    sleeper.sleepNanos(baseDelayNanos + jitterNanos);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    // Record failure for THIS transaction and stop the worker.
                    success = false;
                    stats.add(Boolean.FALSE);
                    return;
                }
            }

            // Exactly one stats entry per transaction
            stats.add(success);
            if (success) {
                result.incrementAndGet();
            }
        }
    }
}
